main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala

```diff
@@ -36,168 +36,206 @@ object SimplifiedAst {
   sealed trait Expression {
     def tpe: Type
 
+    def purity: Purity
+
     def loc: SourceLocation
   }
 
   object Expression {
 
     case class Unit(loc: SourceLocation) extends SimplifiedAst.Expression {
       def tpe: Type = Type.Unit
+      def purity: Purity = Pure
     }
 
-    case class Null(tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression
+    case class Null(tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression {
+      def purity: Purity = Pure
+    }
 
     case class True(loc: SourceLocation) extends SimplifiedAst.Expression {
       def tpe: Type = Type.Bool
+      def purity: Purity = Pure
     }
 
     case class False(loc: SourceLocation) extends SimplifiedAst.Expression {
       def tpe: Type = Type.Bool
+      def purity: Purity = Pure
     }
 
     case class Char(lit: scala.Char, loc: SourceLocation) extends SimplifiedAst.Expression {
       def tpe: Type = Type.Char
+      def purity: Purity = Pure
     }
 
     case class Float32(lit: scala.Float, loc: SourceLocation) extends SimplifiedAst.Expression {
       def tpe: Type = Type.Float32
+      def purity: Purity = Pure
     }
 
     case class Float64(lit: scala.Double, loc: SourceLocation) extends SimplifiedAst.Expression {
       def tpe: Type = Type.Float64
+      def purity: Purity = Pure
     }
 
     case class Int8(lit: scala.Byte, loc: SourceLocation) extends SimplifiedAst.Expression {
       def tpe: Type = Type.Int8
+      def purity: Purity = Pure
     }
 
     case class Int16(lit: scala.Short, loc: SourceLocation) extends SimplifiedAst.Expression {
       def tpe: Type = Type.Int16
+      def purity: Purity = Pure
     }
 
     case class Int32(lit: scala.Int, loc: SourceLocation) extends SimplifiedAst.Expression {
       def tpe: Type = Type.Int32
+      def purity: Purity = Pure
     }
 
     case class Int64(lit: scala.Long, loc: SourceLocation) extends SimplifiedAst.Expression {
       def tpe: Type = Type.Int64
+      def purity: Purity = Pure
     }
 
     case class BigInt(lit: java.math.BigInteger, loc: SourceLocation) extends SimplifiedAst.Expression {
       def tpe: Type = Type.BigInt
+      def purity: Purity = Pure
     }
 
     case class Str(lit: java.lang.String, loc: SourceLocation) extends SimplifiedAst.Expression {
       def tpe: Type = Type.Str
+      def purity: Purity = Pure
     }
 
-    case class Var(sym: Symbol.VarSym, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression
+    case class Var(sym: Symbol.VarSym, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression {
+      def purity: Purity = Pure
+    }
 
-    case class Def(sym: Symbol.DefnSym, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression
+    case class Def(sym: Symbol.DefnSym, tpe: Type, purity: Purity, loc: SourceLocation) extends SimplifiedAst.Expression
```

Comment:
Are there cases when this can be impure?

Reply:
This is always pure.